### PR TITLE
Transformations: Selectively apply transformation to queries

### PR DESCRIPTION
**Pull request by ryantxu**
Currently, transformations are applied to all results.  Usually this is OK, but sometimes we need to modify the results from a specific frame/query rather than uniformly across everything.

This adds an option to filter the frames before applying.  

TODO (if this seems like a good idea):
- [x] tests
- [x] output field order?  currently always appended, but that might be weird

### CODE REVIEW

**Comment by ryantxu** [comment]
| Key | Value |
|--|--|
| Panel | table @ 9.4.0-pre |
| Grafana | 9.4.0-pre (29c8ce12a0) // Open Source |
<details><summary>Panel debug snapshot dashboard</summary>

```json
{
  "panels": [
    {
      "id": 2,
      "gridPos": {
        "h": 13,
        "w": 15,
        "x": 0,
        "y": 0
      },
      "type": "table",
      "title": "Reproduced with embedded data",
      "datasource": {
        "type": "grafana",
        "uid": "grafana"
      },
      "pluginVersion": "9.4.0-pre",
      "fieldConfig": {
        "defaults": {
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          },
          "color": {
            "mode": "thresholds"
          }
        },
        "overrides": []
      },
      "options": {
        "showHeader": true,
        "footer": {
          "show": false,
          "reducer": [
            "sum"
          ],
          "countRows": false,
          "fields": ""
        },
        "frameIndex": 0
      },
      "targets": [
        {
          "refId": "A",
          "datasource": {
            "type": "grafana",
            "uid": "grafana"
          },
          "queryType": "snapshot",
          "snapshot": [
            {
              "schema": {
                "refId": "A",
                "fields": [
                  {
                    "name": "One",
                    "type": "number",
                    "typeInfo": {
                      "frame": "int64",
                      "nullable": true
                    },
                    "config": {}
                  },
                  {
                    "name": "Two",
                    "type": "number",
                    "typeInfo": {
                      "frame": "int64",
                      "nullable": true
                    },
                    "config": {}
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    1,
                    3
                  ],
                  [
                    2,
                    4
                  ]
                ]
              }
            },
            {
              "schema": {
                "refId": "B",
                "fields": [
                  {
                    "name": "BBB",
                    "type": "number",
                    "typeInfo": {
                      "frame": "int64",
                      "nullable": true
                    },
                    "config": {}
                  }
                ]
              },
              "data": {
                "values": [
                  [
                    4,
                    5
                  ]
                ]
              }
            }
          ]
        }
      ],
      "transformations": [
        {
          "id": "calculateField",
          "options": {},
          "filter": {
            "id": "byRefId",
            "options": "B"
          }
        }
      ]
    },
    {
      "gridPos": {
        "h": 7,
        "w": 9,
        "x": 15,
        "y": 0
      },
      "id": 5,
      "options": {
        "content": "<table width=\"100%\">\n    <tr>\n      <th width=\"2%\">Panel</th>\n      <td >table @ 9.4.0-pre</td>\n    </tr>\n    <tr>\n      <th>Queries</th>\n      <td>A[testdata], B[testdata]</td>\n    </tr>\n    <tr>\n      <th>Transforms (1)</th>\n      <td>calculateField</td>\n  </tr>\n    <tr><th>Data</th><td> 2 frames, 3 fields, 4 rows </td></tr>\n    \n    <tr>\n      <th>Grafana</th>\n      <td>9.4.0-pre (29c8ce12a0) // Open Source</td>\n    </tr>\n  </table>",
        "mode": "html"
      },
      "title": "Debug info",
      "type": "text"
    },
    {
      "id": 6,
      "title": "Original Panel JSON",
      "type": "text",
      "gridPos": {
        "h": 13,
        "w": 9,
        "x": 15,
        "y": 7
      },
      "options": {
        "content": "{\n  \"id\": 2,\n  \"gridPos\": {\n    \"h\": 9,\n    \"w\": 12,\n    \"x\": 0,\n    \"y\": 0\n  },\n  \"type\": \"table\",\n  \"title\": \"Panel Title\",\n  \"datasource\": {\n    \"uid\": \"PD8C576611E62080A\",\n    \"type\": \"testdata\"\n  },\n  \"pluginVersion\": \"9.4.0-pre\",\n  \"fieldConfig\": {\n    \"defaults\": {\n      \"custom\": {\n        \"align\": \"auto\",\n        \"cellOptions\": {\n          \"type\": \"auto\"\n        },\n        \"inspect\": false\n      },\n      \"mappings\": [],\n      \"thresholds\": {\n        \"mode\": \"absolute\",\n        \"steps\": [\n          {\n            \"color\": \"green\",\n            \"value\": null\n          },\n          {\n            \"color\": \"red\",\n            \"value\": 80\n          }\n        ]\n      },\n      \"color\": {\n        \"mode\": \"thresholds\"\n      }\n    },\n    \"overrides\": []\n  },\n  \"options\": {\n    \"showHeader\": true,\n    \"footer\": {\n      \"show\": false,\n      \"reducer\": [\n        \"sum\"\n      ],\n      \"countRows\": false,\n      \"fields\": \"\"\n    },\n    \"frameIndex\": 0\n  },\n  \"targets\": [\n    {\n      \"scenarioId\": \"csv_content\",\n      \"refId\": \"A\",\n      \"datasource\": {\n        \"type\": \"testdata\",\n        \"uid\": \"PD8C576611E62080A\"\n      },\n      \"csvContent\": \"One,Two\\n1,2\\n3,4\"\n    },\n    {\n      \"scenarioId\": \"csv_content\",\n      \"refId\": \"B\",\n      \"datasource\": {\n        \"type\": \"testdata\",\n        \"uid\": \"PD8C576611E62080A\"\n      },\n      \"csvContent\": \"BBB\\n4\\n5\",\n      \"hide\": false\n    }\n  ],\n  \"transformations\": [\n    {\n      \"id\": \"calculateField\",\n      \"options\": {},\n      \"filter\": {\n        \"id\": \"byRefId\",\n        \"options\": \"B\"\n      }\n    }\n  ]\n}",
        "mode": "code",
        "code": {
          "language": "json",
          "showLineNumbers": true,
          "showMiniMap": true
        }
      }
    },
    {
      "id": 3,
      "title": "Data from panel above (after transformations)",
      "type": "table",
      "datasource": {
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "gridPos": {
        "h": 7,
        "w": 15,
        "x": 0,
        "y": 13
      },
      "options": {
        "showTypeIcons": true
      },
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 2,
          "withTransforms": true,
          "refId": "A"
        }
      ]
    },
    {
      "id": 100,
      "title": "Data (before transformations)",
      "type": "table",
      "datasource": {
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "gridPos": {
        "h": 7,
        "w": 24,
        "x": 0,
        "y": 13
      },
      "options": {
        "showTypeIcons": true
      },
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 2,
          "withTransforms": false,
          "refId": "A"
        }
      ]
    }
  ],
  "schemaVersion": 37,
  "title": "Debug: Panel Title // 2023-01-18 23:00:39",
  "tags": [
    "debug",
    "debug-table"
  ],
  "time": {
    "from": "2023-01-19T01:00:39.665Z",
    "to": "2023-01-19T07:00:39.665Z"
  }
}
```
</details>

**Comment by ryantxu** [comment]
Started poking at this because of: https://github.com/grafana/iot-sitewise-datasource/issues/145#issuecomment-1331464319

File: public/app/features/dashboard/components/TransformationsEditor/TransformationFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  const styles = useStyles2(getStyles);
+  const context = useMemo(() => {
+    // eslint-disable-next-line
+    return { data } as StandardEditorContext<unknown>;
+  }, [data]);
+
+  return (
+    <div className={styles.wrapper}>
+      <h5>Apply tranformation to</h5>
+      <FrameSelectionEditor
```
Comment (ryantxu) [inline public/app/features/dashboard/components/TransformationsEditor/TransformationFilter.tsx:31]: This component is used in gemap to pick a single frame.  This is sufficient to get started, but note that the config object is any frame matcher, so we will eventually need something (in both places) that is more robust

File: packages/grafana-data/src/transformations/matchers/ids.ts [hunk 2]
```diff
@@ -35,7 +35,6 @@ export enum FrameMatcherID {
   byName = 'byName',
   byRefId = 'byRefId',
   byIndex = 'byIndex',
-  byLabel = 'byLabel',
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/matchers/ids.ts:38]: This is why levieate fails... but it is not a valid value so better I think

**Comment by gtk-grafana** [comment]
This looks good but I'm having trouble reproducing this locally, could we add a gdev dashboard demoing this functionality? (also a debug snapshot would be appreciated)

**Review by gtk-grafana — COMMENTED** [review]
Code looks good, but when testing I'm seeing some conflicting behavior with existing transformations, at least those where you can already select a refId.
Take a look at "Filter data by query", and you can see that it's functionality stops working when the filter is selected. For this specific transformation, I might suggest removing the filter icon.

Also Extract fields seems to have a dropdown to select which query you want to target, which is impacted by this change as well, maybe we just need to remove this functionality for every transform that already has UI for picking which query the transform is applied to?

**Comment by ryantxu** [comment]
@gtk-grafana -- I updated the UI so it now makes the organize use case much more clear.  Looks like partition by value has the same issue also:

File: public/app/features/transformers/editors/FilterByRefIdTransformerEditor.tsx [hunk 3]
```diff
@@ -103,28 +103,36 @@ export class FilterByRefIdTransformerEditor extends React.PureComponent<
...
-                    this.onFieldToggle(o.refId);
-                  }}
-                  label={label}
-                  selected={isSelected}
-                />
-              );
-            })}
-          </HorizontalGroup>
+      <>
+        {input.length <= 1 && (
```
[the remark this answers is not in the document]
Comment (gtk-grafana) [reply public/app/features/transformers/editors/FilterByRefIdTransformerEditor.tsx:109]: Yeah I didn't initially understand the functionality

**Review by gtk-grafana — APPROVED** [review]
This is looking good and working for me locally. Thanks for taking the time to add this, really cool stuff!!

**Comment by samjewell** [comment]
I love this feature - it makes Transformations far more powerful.

Today I found myself wishing I could **multi-select** which queries I wanted to filter down to. I had three queries, and I wanted to choose two of them when making an OUTER JOIN transform. Because I wanted to retain the 3rd query, and join it in much later on after many other joins had been applied. I thought I'd drop that idea here, and see what you thought? Thanks!

**Pull request merged**